ECv2 support

kse-clearhaus opened the pull request.

kse-clearhaus committed:
Initial ECv2 support

kse-clearhaus committed:
bugfix/update gem version

kse-clearhaus committed:
Bump gem to version 0.9.0

kse-clearhaus committed:
Omit 'merchant:' prefix in merchant_id

kse-clearhaus committed:
bugfix merchant signing string

mt-clearhaus commented on lib/aliquot-pay.rb:
```diff
@@ -72,36 +72,63 @@ def self.four_byte_length(str)
+  def self.generate_signature(*args)
+    args.map do |s|
+      four_byte_length(s) + s
+    end.join('')
```
Space as separator is default.

ct-clearhaus replied:
*Empty string*  as separator is default 🙂

mt-clearhaus replied:
I'm sorry... I thought "empty string" but wrote "space"...

mt-clearhaus commented on lib/aliquot-pay.rb:
```diff
@@ -72,36 +72,63 @@ def self.four_byte_length(str)
+  def self.generate_token_ecv2(payment, signing_key, intermediate_key, recipient,
+                               signed_message: nil, expire_time:  "#{Time.now.to_i + 3600}000")
+    cipher = OpenSSL::Cipher::AES256.new(:CTR)
+    signed_message ||= JSON.unparse(encrypt(JSON.unparse(payment), recipient, cipher))
```
Can you explain the two instances of JSON unparsing here? I am a bit confused :thinking:

ct-clearhaus replied:
The inner `payment` is `#to_json`'ed (`JSON.unparse`'d) before it's encrypted ...

kse-clearhaus replied:
Exactly. The `payment` is added as serialized JSON, this is done because it is what is verified by `tag`, a MAC.
This results in a signed message, which is also a serialized JSON string in the final token, since it is signed by Google.

mt-clearhaus commented on lib/aliquot-pay/util.rb:
```diff
@@ -11,18 +11,18 @@ def self.generate_shared_secret(private_key, public_key)
       if OpenSSL.const_defined?(:KDF) && OpenSSL::KDF.respond_to?(:hkdf)
         h = OpenSSL::Digest::SHA256.new
-        hbytes = OpenSSL::KDF.hkdf(input_keying_material, hash: h, salt: '', length: 32, info: info)
+        hbytes = OpenSSL::KDF.hkdf(input_keying_material, hash: h, salt: '', length: length * 2, info: info)
```
So 64 is correct and 32 was incorrect? Did it work before?

kse-clearhaus replied:
Both 64 and 32 are correct.
For `ECv1` you need to generate 32 bytes of key material, 128 bits for AES key and 128 bits for MAC key.
For `ECv2`you need 64 bytes, 256 bits for each.

kse-clearhaus replied:
Since `ECv2` has superseded `ECv1` the required 64 bytes of keys is now default.

ct-clearhaus replied:
With v1 I suspect?

kse-clearhaus replied:
What do you mean ct?
Yes, it worked before with `ECv1`.

mt-clearhaus commented on lib/aliquot-pay/util.rb:
```diff
@@ -11,18 +11,18 @@ def self.generate_shared_secret(private_key, public_key)
-        aes_key: hbytes[0..15],
-        mac_key: hbytes[16..32],
+        aes_key: hbytes[0..length - 1],
+        mac_key: hbytes[length..2 * length],
```
Not `length..(2 * length) - 1` ? I.e. 32..63.

mt-clearhaus replied:
Alternatively, `length..-1`

kse-clearhaus replied:
Yes, it is. luckily it still performs as it should. Now I'm considering it, I prefer.
```ruby
        aes_key: hbytes[0, length],
        mac_key: hbytes[length, length],
```

ct-clearhaus commented on lib/aliquot-pay.rb:
```diff
@@ -72,36 +72,63 @@ def self.four_byte_length(str)
-    recipient_id     = DEFAULTS[:merchant_id],
-    sender_id        = DEFAULTS[:info],
-    protocol_version = 'ECv1'
+    recipient_id:     "merchant:#{DEFAULTS[:merchant_id]}",
```
Shouldn't this be a parameter (that defaults to `DEFAULTS[:merchant_id]`)?

(Likewise for `DEFAULTS[:info]` below.)

ct-clearhaus replied:
Strike that. It's exactly what it is 🤦‍♂️

ct-clearhaus commented on lib/aliquot-pay.rb:
```diff
@@ -72,36 +72,63 @@ def self.four_byte_length(str)
-    recipient_id     = DEFAULTS[:merchant_id],
-    sender_id        = DEFAULTS[:info],
-    protocol_version = 'ECv1'
+    recipient_id:     "merchant:#{DEFAULTS[:merchant_id]}",
```
Uhm, but if the `recipient_id` does always start with `"merchant:"` then ... couldn't we "hide" this "deeper" so the user calling `.signature_string` will not have to prepend the merchant ID with `"merchant:"`? Perhaps it's nicer to do it like you do — for debugging and spec'ing purposes?

kse-clearhaus replied:
Hmm. I see no benefit in doing it like I do it. As in, we don't use it right now.
I agree that it should be `merchant_id` as a parameter rather than `recipient_id`.

ct-clearhaus replied:
If convenient for debugging/spec'ing, you could allow both 🙂

ct-clearhaus commented on lib/aliquot-pay/util.rb:
```diff
@@ -11,18 +11,18 @@ def self.generate_shared_secret(private_key, public_key)
     end
 
-    def self.derive_keys(ephemeral_public_key, shared_secret, info)
+    def self.derive_keys(ephemeral_public_key, shared_secret, info, length: 32)
```
Thought: What about supplying the "EC version" rather than `length`? I think that could perhaps be more user friendly?

kse-clearhaus replied:
Hmm. Sure, that makes sense. :+1:

kse-clearhaus commented:
I sort of feel like this library is badly written for generating tokens.
In my opinion a more object oriented construction would be better. The real question is whether I want to spend time on rewriting it.

kse-clearhaus committed:
Implement review comments

- `JSON.unparse` => `#to_json`
- Derive cipher from protocol_version instead of passing cipher object
- Fix indexing of MAC/AES keys

kse-clearhaus committed:
bump gem version to 0.12.0

kse-clearhaus commented:
@ct-clearhaus @mt-clearhaus Any more comments?

mt-clearhaus reviewed (approved):
LGTM :+1:
